src/main/scala/mesosphere/marathon/core/heartbeat/HeartbeatActor.scala

```diff
@@ -0,0 +1,97 @@
+package mesosphere.marathon.core.heartbeat
+
+import akka.actor._
+import scala.concurrent.duration._
+
+// HeartbeatActor monitors the heartbeat of some process and executes handlers for various conditions.
+// If an expected heartbeat is missed then execute an `onSkip` handler.
+// If X number of subsequent heartbeats are missed then execute an `onFailure` handler and become inactive.
+// Upon creation the actor is in an inactive state and must be sent a MessageActivate message to activate.
+// Once activated the actor will monitor for MessagePulse messages (these are the heartbeats).
+// The actor may be forcefully deactivated by sending it an MessageDeactivate message.
+class HeartbeatActor(config: Heartbeat.Config) extends LoggingFSM[HeartbeatInternal.State, HeartbeatInternal.Data] {
+  import Heartbeat._
+  import HeartbeatInternal._
+
+  startWith(StateInactive, DataNone)
+
+  when(StateInactive) {
+    case Event(MessageActivate(reactor, token), DataNone) =>
+      log.debug("heartbeat activated")
+      goto(StateActive) using DataActive(reactor, token)
+    case _ =>
+      stay // swallow all other event types
+  }
+
+  when(StateActive, stateTimeout = config.heartbeatTimeout) {
+    case Event(MessagePulse, data: DataActive) =>
+      stay using data.copy(missed = 0)
+
+    case Event(StateTimeout, data: DataActive) =>
+      if (data.missed + 1 >= config.missedHeartbeatsThreshold) {
+        data.reactor.onFailure
+        goto(StateInactive) using DataNone
+      } else {
+        data.reactor.onSkip
+        stay using data.copy(missed = data.missed + 1)
+      }
+
+    case Event(MessageDeactivate(token), data: DataActive) =>
+      // only deactivate if token == data.sessionToken
+      if (token.eq(data.sessionToken)) {
+        log.debug("heartbeat deactivated")
+        goto(StateInactive) using DataNone
+      } else {
+        stay
+      }
+
+    case Event(MessageActivate(newReactor, newToken), data: DataActive) =>
+      log.debug("heartbeat re-activated")
+      stay using DataActive(reactor = newReactor, sessionToken = newToken)
+  }
+
+  whenUnhandled{
+    case Event(e, d) =>
+      log.warning("unhandled event {} in state {}/{}", e, stateName, d)
+      stay
+  }
+
+  log.debug("starting heartbeat actor")
+
+  initialize()
+}
+
+object Heartbeat {
+  case class Config(
+    heartbeatTimeout: FiniteDuration,
+    missedHeartbeatsThreshold: Int)
+
+  sealed trait Message
+  case object MessagePulse extends Message
+  case class MessageDeactivate(sessionToken: AnyRef) extends Message
+  case class MessageActivate(reactor: Reactor, sessionToken: AnyRef) extends Message
+
+  trait Reactor {
+    def onSkip(): Unit
+    def onFailure(): Unit
+  }
+
+  def props(config: Config): Props = Props(classOf[HeartbeatActor], config)
+}
+
+protected[heartbeat] object HeartbeatInternal {
+  import Heartbeat._
+
+  sealed trait State
+  case object StateInactive extends State
```

Comment:
yes, you did comment on this prior. I tried it and found the code harder to
read. The convention I used carries over from golang and IMHO makes it
easier to differentiate among enum'd types. So, I left it as is. I'm happy
to revise if people find this confusing.

TASK_STARTING is not a good name, it's already encapsulated by TaskState.
so TaskState.STARTING would be sufficient for context and to reduce
stutter. Oh well, that's inherited from the Mesos API.

FWIW in this PR StateInactive and MessageDeactivate do not stutter.



On Fri, Jul 15, 2016 at 2:40 PM, Matthias Eichstedt <
notifications@github.com> wrote:

> In src/main/scala/mesosphere/marathon/core/heartbeat/HeartbeatActor.scala
> <https://github.com/mesosphere/marathon/pull/4069#discussion_r71021638>:
>
> > +  case class MessageDeactivate(sessionToken: AnyRef) extends Message
> > +  case class MessageActivate(reactor: Reactor, sessionToken: AnyRef) extends Message
> > +
> > +  trait Reactor {
> > +    def onSkip(): Unit
> > +    def onFailure(): Unit
> > +  }
> > +
> > +  def props(config: Config): Props = Props(classOf[HeartbeatActor], config)
> > +}
> > +
> > +protected[heartbeat] object HeartbeatInternal {
> > +  import Heartbeat._
> > +
> > +  sealed trait State
> > +  case object StateInactive extends State
>
> Iirc I've added a note about this before which probably wasn't obvious:
> imho Inactive would be sufficient to denote a state, while Deactivate
> would be sufficient to denote a message – I don't think we need a prefix,
> just as Starting seems like a catchier task state to me than TASK_STARTING
> .
>
> We don't have a clear convention here and I don't feel too strong about
> this. I'd be ok to leave it like that and discuss mid of August ;)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mesosphere/marathon/pull/4069/files/718c11edd708bc282deea7cca641f663168b5c7c#r71021638>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACPVLFdyPcD8CR-54REfIqgt47wFRCNgks5qV9QvgaJpZM4JGoyy>
> .
>